### PR TITLE
Implement metrics/tracing on top of OpenTelemetry

### PR DESCRIPTION
**Pull request by Pliner**

### CODE REVIEW

**Review by slawwan — APPROVED** [review]

**Review by slawwan — APPROVED** [review]

File: asyncpg_listen/listener.py [hunk 1]
```diff
@@ -169,3 +178,32 @@ def _push(_: Any, __: Any, channel: Any, payload: Any) -> None:
...
+            name = f"Notification #{notification.channel}"
+        else:
+            raise TypeError(f"Unexpected notification type: {type(notification)}")
+
+        started_at = time.time_ns()
+
+        with self._tracer.start_as_current_span(
+            name=name,
+            kind=opentelemetry.trace.SpanKind.INTERNAL,
+            start_time=started_at,
```
Comment (slawwan) [inline asyncpg_listen/listener.py:202]: started_at -> start_time ? like arg name?

File: asyncpg_listen/listener.py [hunk 2]
```diff
@@ -169,3 +178,32 @@ def _push(_: Any, __: Any, channel: Any, payload: Any) -> None:
...
+            self._tracer = opentelemetry.trace.get_tracer(__package__)
+        if self._meter is None:
+            self._meter = opentelemetry.metrics.get_meter(__package__)
+        if self._notification_histogram is None:
+            self._notification_histogram = self._meter.create_histogram("asyncpg_listener_notification")
+
+        if isinstance(notification, Timeout):
+            name = f"Notification timeout #{notification.channel}"
+        elif isinstance(notification, Notification):
+            name = f"Notification #{notification.channel}"
```
Comment (slawwan) [inline asyncpg_listen/listener.py:193]: what if move this iff-else block to separate function `get_notification_name(notification) -> str` ?

File: asyncpg_listen/listener.py [hunk 3]
```diff
@@ -169,3 +178,32 @@ def _push(_: Any, __: Any, channel: Any, payload: Any) -> None:
...
 
         return _push
+
+    async def _process_notification(self, handler: NotificationHandler, notification: NotificationOrTimeout) -> None:
+        if self._tracer is None:
+            self._tracer = opentelemetry.trace.get_tracer(__package__)
+        if self._meter is None:
+            self._meter = opentelemetry.metrics.get_meter(__package__)
+        if self._notification_histogram is None:
+            self._notification_histogram = self._meter.create_histogram("asyncpg_listener_notification")
```
Comment (slawwan) [inline asyncpg_listen/listener.py:188]: what if  move this three `iff` blocks to funciton like `_ensure_setup()  -> NamedTuple(tracecr, meter. notification)` or something like this? or make properties with backing fields?

File: asyncpg_listen/listener.py (hunk 1 above)
Comment (Pliner) [reply asyncpg_listen/listener.py]: Fixed

File: asyncpg_listen/listener.py (hunk 2 above)
Comment (Pliner) [reply asyncpg_listen/listener.py]: Fixed

**Pull request merged**